source/extensions/watchdog/profile_action/profile_action.cc

```diff
@@ -0,0 +1,115 @@
+#include "extensions/watchdog/profile_action/profile_action.h"
+
+#include <chrono>
+
+#include "envoy/thread/thread.h"
+
+#include "common/profiler/profiler.h"
+#include "common/protobuf/utility.h"
+
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Watchdog {
+namespace ProfileAction {
+namespace {
+static constexpr uint64_t DefaultMaxProfilePerTid = 10;
+
+std::string generateProfileFilePath(const std::string& directory, const SystemTime& now) {
+  auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
+  if (absl::EndsWith(directory, "/")) {
+    return absl::StrFormat("%s%s.%d", directory, "ProfileAction", timestamp);
+  }
+  return absl::StrFormat("%s/%s.%d", directory, "ProfileAction", timestamp);
+}
+} // namespace
+
+ProfileAction::ProfileAction(
+    envoy::extensions::watchdog::profile_action::v3alpha::ProfileActionConfig& config,
+    Server::Configuration::GuardDogActionFactoryContext& context)
+    : path_(config.profile_path()), running_profile_(false),
+      max_profiles_per_tid_(config.max_profiles_per_thread() ? config.max_profiles_per_thread()
+                                                             : DefaultMaxProfilePerTid),
+      profiles_started_(0), duration_(std::chrono::milliseconds(
+                                PROTOBUF_GET_MS_OR_DEFAULT(config, profile_duration, 5000))),
+      context_(context) {}
+
+void ProfileAction::run(
+    envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent /*event*/,
+    std::vector<std::pair<Thread::ThreadId, MonotonicTime>> thread_ltt_pairs,
+    MonotonicTime /*now*/) {
+  if (running_profile_) {
+    return;
+  }
+
+  // Check if there's a tid that justifies profiling
+  auto trigger_tid = getTidTriggeringProfile(thread_ltt_pairs);
+  if (!trigger_tid.has_value()) {
+    ENVOY_LOG_MISC(warn, "None of the provide tids justify profiling");
+    return;
+  }
+
+  auto& fs = context_.api_.fileSystem();
+  if (!fs.directoryExists(path_)) {
+    ENVOY_LOG_MISC(error, "Directory path {} doesn't exists.", path_);
+    return;
+  }
+
+  // Generate file path for output and try to profile
+  const auto& profile_filename =
+      generateProfileFilePath(path_, context_.api_.timeSource().systemTime());
+
+  if (!Profiler::Cpu::profilerEnabled()) {
+    if (Profiler::Cpu::startProfiler(profile_filename)) {
+      // Update state
+      running_profile_ = true;
+      ++profiles_started_;
+      tid_to_profile_count_[*trigger_tid] += 1;
+
+      // Schedule callback to stop
+      timer_cb_ = context_.dispatcher_.createTimer([this, profile_filename] {
+        if (Profiler::Cpu::profilerEnabled()) {
+          Profiler::Cpu::stopProfiler();
+          running_profile_ = false;
+          timer_cb_->disableTimer();
```

Comment:
I don't understand this call. It looks like the timer is only enabled here, and timers are always one-shot, so the timer should be disabled after this callback runs.